docs/assignment3.md

```diff
@@ -1 +1,577 @@
+    </li>
+    <li>
+        Sequence diagram: besides other changes that had to be made in order to adjust the diagram to the new game implementation, we got rid of the first alt statement, since it did not add any special behavior. 
+    </li>
+    <li>
+        unused variables: variables that were not used during assignment 2 were completely removed or finally implemented. Health and moves now have a purpose and are being used. Health was completely removed as it did not match our expectation of the game anymore
+    </li>
+</ul>
+
+
+### Application of design patterns
+Author(s): `Florent Brunet de Rochebrune, Eduardo Lira`
+
+![Design Patterns in Class Diagram](https://i.imgur.com/15e2WKT.png)
+
+| ID                     | DP1 |
+| ---------------------- | ------------------------------------------------------------ |
+| **Design pattern**     | Command Pattern |
+| **Problem**            | The game consists of a server and a client. The client will be sending commands to the server and the server will react with a response. These commands consist of objects which then are directly executed by the server. To add another command, the client and the server will both have to be updated, as the objects are hardcoded. Implementing new methods would be very inflexible and difficult, because it couples the class to a particular request. Adding or removing a command would require re-hardwiring the player client and server. |
+| **Solution**           | We can solve this by creating an Instruction class. This class does not implement a request directly but refers to the server's command interface to perform a request. The client is then independent of how the request is performed. The server will handle the request and execute the command, not the client directly. Creating an Instruction class makes it possible to configure an object with a single request. The Instruction captures all the information needed to trigger an event or action at a later time. The biggest benefit is that nothing is 'hard-wired' anymore. Changing how Instructions are structured can be done by simply changing the Instruction class. |
+| **Intended use**       | This allows the client to send complete Instruction objects (in our case across the network) to be executed on the server. When adding new commands to the server, only the server has to be updated as the client will only be sending requests in a generic textual form, not a specialised object that is directly executed on the server. <br>This is also a security measure, as it removes any client-side processing.|
+| **Constraints**        | There is no defined way of sending these commands (yet). The next design pattern will go more in-depth for that. |
+
+| ID                     | DP2 |
+| ---------------------- | ------------------------------------------------------------ |
+| **Design pattern**     | Interpreter Pattern |
```

Comment:
Also here the interpreter design pattern is not properly implemented. A interpreter function within the Game class is implemented. In the interpreter design pattern, the rules of the grammar are implemented with expression classes which are linked and all implement the interpret() method